enterprise/server/remote_execution/executor/executor.go

```diff
@@ -341,6 +343,34 @@ func (s *Executor) ExecuteTaskAndStreamResults(ctx context.Context, st *repb.Sch
 		log.CtxInfof(ctx, "Task finished cleanly.")
 		finishedCleanly = true
 	}
+
+	// If the runner has exceeded 90% of memory or disk usage, don't try to recycle
+	// it, because that may cause failures if it's reused
+	usedMemoryBytes := cmdResult.UsageStats.MemoryBytes
+	totalMemoryBytes := st.GetSchedulingMetadata().GetTaskSize().GetEstimatedMemoryBytes()
+	maxedOutStr := ""
+	if usedMemoryBytes >= int64(float64(totalMemoryBytes)*.9) {
+		maxedOutStr += fmt.Sprintf("%d/%d B memory used", usedMemoryBytes, totalMemoryBytes)
+	}
+	for _, fsUsage := range cmdResult.UsageStats.GetPeakFileSystemUsage() {
+		if float64(fsUsage.UsedBytes)/float64(fsUsage.TotalBytes) >= .9 {
+			maxedOutStr += fmt.Sprintf(" %d/%d B disk used for %s", fsUsage.UsedBytes, fsUsage.TotalBytes, fsUsage.GetSource())
+		}
+	}
+
+	if maxedOutStr != "" {
+		finishedCleanly = false
+
+		errStr := fmt.Sprintf("%v runner exceeded 90%% of memory or disk usage, not recycling: %s", r.GetIsolationType(), maxedOutStr)
+		if cr, ok := r.(*runner.CommandRunner); ok {
+			if fc, ok := cr.Container.Delegate.(*firecracker.FirecrackerContainer); ok {
+				errStr += fmt.Sprintf("\nSnapshot debug key: %s", fc.SnapshotDebugString(ctx))
+			}
+		}
+		alert.UnexpectedEvent("runner_maxed_out", errStr)
+		log.Errorf(errStr)
+	}
```

Comment:
Can we use the existing `doNotRecycle` bool in the `commandRunner` and move this logic to `Run()` ? I think that would make the PR a lot smaller

Reply:
Good idea - changed